udpsession.go

```diff
@@ -0,0 +1,364 @@
+package session
+
+import (
+	"log"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/nknorg/nkn-tuna-session/pb"
+	"github.com/nknorg/tuna"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// endpoint type, listener or dialer
+	EndpointListener = 1
+	EndpointDialer   = 2
+)
+
+// Wrap of UDPAddr, including session information
+// implement of net.Addr interface
+type udpSessionAddr struct {
+	remoteAddr string
+	sessionID  []byte
+	udpAddr    *net.UDPAddr
+}
+
+func NewUdpSessionAddr(remoteAddr string, sessionID []byte, udpAddr *net.UDPAddr) *udpSessionAddr {
+	return &udpSessionAddr{remoteAddr: remoteAddr, sessionID: sessionID, udpAddr: udpAddr}
+}
+func (usa *udpSessionAddr) Network() string {
+	return "Tuna Session UDP"
+}
+func (usa *udpSessionAddr) String() string {
+	return sessionKey(usa.remoteAddr, usa.sessionID)
+}
+
+// UDP session, it is a UDP listener or dialer endpoint for reading and writing udp data.
+// After tuna session set up UDP connection, it return UpdSession to applications.
+// Applications use this end point to read and write data.
+// If there is a node break down between UDP listener and dialer, udp session will reconenct automatically
+// without interrupting application usage.
+type UdpSession struct {
+	endpointType int // EndpointDialer or EndpointListener
+	udpConn      *tuna.EncryptUDPConn
+	ts           *TunaSessionClient // Reference of tuna session
+
+	// dialer only
+	tcpConn *Conn           // TCP *Conn, dialer only
+	peer    *udpSessionAddr // dialer only
+
+	// listener only
+	sync.RWMutex
+	sessKeyToSessAddr map[string]*udpSessionAddr // listener only, udpSessionAddr.String() to *udpSessionAddr
+	udpAddrToSessAddr map[string]*udpSessionAddr // listener only, net.UDPAddr.String() to *udpSessionAddr
+	tcpConns          map[string]*Conn           // listen only, session key to *Conn
+}
+
+func newUdpSession(ts *TunaSessionClient, udpConn *tuna.EncryptUDPConn, endpointType int, peerAddr *udpSessionAddr) *UdpSession {
+	us := &UdpSession{ts: ts, udpConn: udpConn, endpointType: endpointType, peer: peerAddr}
+	if endpointType == EndpointListener {
+		us.sessKeyToSessAddr = make(map[string]*udpSessionAddr)
+		us.udpAddrToSessAddr = make(map[string]*udpSessionAddr)
+		if peerAddr != nil {
+			us.sessKeyToSessAddr[peerAddr.String()] = peerAddr
+			us.udpAddrToSessAddr[peerAddr.udpAddr.String()] = peerAddr
+		}
+		us.tcpConns = make(map[string]*Conn)
+	}
+
+	return us
+}
+
+// ReadFrom, implement of standard library net.UDPConn.ReadFrom
+func (us *UdpSession) ReadFrom(b []byte) (int, net.Addr, error) {
+	for {
+		buf := make([]byte, tuna.MaxUDPBufferSize)
+		n, udpAddr, err := us.udpConn.ReadFromUDP(buf)
+		if err != nil {
+			return 0, nil, err
+		}
+
+		header := buf[0]
+		switch header {
+		case byte(pb.HeaderType_USER):
+			us.RLock()
+			addr, ok := us.udpAddrToSessAddr[udpAddr.String()]
+			us.RUnlock()
+			if ok {
+				copy(b, buf[1:n]) // remove header, and return to user application

```

Comment:
Need to decrypt message. The easiest way is to call AddCodec once you know the sender

Reply:
Yes, added.